x/datadeal/keeper/deal.go

```diff
@@ -0,0 +1,144 @@
+package keeper
+
+import (
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	gogotypes "github.com/gogo/protobuf/types"
+	"github.com/medibloc/panacea-core/v2/x/datadeal/types"
+)
+
+func (k Keeper) CreateDeal(ctx sdk.Context, buyerAddress sdk.AccAddress, msg *types.MsgCreateDeal) (uint64, error) {
+
+	dealID, err := k.GetNextDealNumberAndIncrement(ctx)
+	if err != nil {
+		return 0, sdkerrors.Wrapf(err, "failed to get next deal num")
+	}
+
+	newDeal := types.NewDeal(dealID, msg)
+
+	coins := sdk.NewCoins(*msg.Budget)
+
+	dealAddress, err := sdk.AccAddressFromBech32(newDeal.Address)
+	if err != nil {
+		return 0, err
+	}
+
+	acc := k.accountKeeper.GetAccount(ctx, dealAddress)
+	if acc != nil {
+		return 0, sdkerrors.Wrapf(types.ErrDealAlreadyExist, "deal %d already exist", dealID)
+	}
+
+	acc = k.accountKeeper.NewAccount(ctx, authtypes.NewModuleAccount(
+		authtypes.NewBaseAccountWithAddress(
+			dealAddress,
+		),
+		strconv.FormatUint(newDeal.Id, 10)),
+	)
+	k.accountKeeper.SetAccount(ctx, acc)
+
+	if err = k.bankKeeper.SendCoins(ctx, buyerAddress, dealAddress, coins); err != nil {
+		return 0, sdkerrors.Wrapf(err, "Failed to send coins to deal account")
+	}
+
+	if err = k.SetDeal(ctx, *newDeal); err != nil {
+		return 0, err
+	}
+
+	return newDeal.Id, nil
+}
+
+func (k Keeper) SetNextDealNumber(ctx sdk.Context, dealNumber uint64) error {
+	store := ctx.KVStore(k.storeKey)
+	bz, err := k.cdc.MarshalLengthPrefixed(&gogotypes.UInt64Value{Value: dealNumber})
+	if err != nil {
+		return sdkerrors.Wrapf(err, "Failed to set next deal number")
+	}
+	store.Set(types.KeyDealNextNumber, bz)
+	return nil
+}
+
+func (k Keeper) GetNextDealNumber(ctx sdk.Context) (uint64, error) {
+	var dealNumber uint64
+	store := ctx.KVStore(k.storeKey)
+
+	if !store.Has(types.KeyDealNextNumber) {
+		return 0, types.ErrDealNotInitialized
+	}
+
+	bz := store.Get(types.KeyDealNextNumber)
+
+	val := gogotypes.UInt64Value{}
+
+	if err := k.cdc.UnmarshalLengthPrefixed(bz, &val); err != nil {
+		return 0, err
```

Comment:
It seems good to wrap errors with sdkerrors.

Reply:
Thanks, I add sdkerrors.